# 項目：評估和清理英國電商公司銷售數據

## 分析目標

此數據分析的目的是根據市場銷售數據，挖掘暢銷產品，以便制定更有效的市場策略來提升營收。

## 簡介

原始數據集記錄了一家英國線上零售公司在2010年12月1日至2011年12月9日期間的所有交易情況，涵蓋了該公司在全球不同國家和地區的業務數據。該公司主要銷售涵蓋各種場景的禮品，包括但不限於生日禮品、結婚紀念品、聖誕禮品等。該公司的客戶群體主要包括批發商和個人消費者，其中批發商佔據了相當大的比例。

數據每列的含義如下：

- `InvoiceNo`: 發票號碼。6位數，作為交易的唯一標識符。如果這個編號以字母 "C" 開頭，表示該筆交易已被取消。
- `StockCode`: 產品代碼。5位數，作為產品的唯一標識符。
- `Description`: 產品名稱。
- `Quantity`: 產品在交易中的數量。
- `InvoiceDate`: 發票日期和時間，表示交易發生的日期和時間。
- `UnitPrice`: 單價，價格單位為英鎊（£）。
- `CustomerID`: 客戶編號。5位數，作為客戶的唯一標識符。
- `Country`: 國家名稱，表示客戶所居住的國家。

## 讀取數據

In [1]:
import pandas as pd

In [2]:
orginal_data = pd.read_csv("e_commerce.csv")

In [3]:
orginal_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## 評估數據


在這一部分，我將對在上一部分建立的`original_data` 這個 DataFrame所包含的數據進行評估。

評估主要從兩個方面進行：結構和內容，即整齊度和乾淨度。數據的結構性問題指不符合「每列是一個變量，每行是一個觀察值，每個單元格是一個值」這三個標準，數據的內容性問題包括存在遺失數據、重複數據、無效數據等。

## 評估數據整齊度

In [4]:
orginal_data.sample(10)

InvoiceNo StockCode                          Description  Quantity  \
536247    581219     23163                  REGENCY SUGAR TONGS         1   
227734    556905     23080             RED METAL BOX TOP SECRET         1   
225964    556776     23321          SMALL WHITE HEART OF WICKER         1   
446664    574914     23321          SMALL WHITE HEART OF WICKER         3   
111799    545830    47590B          PINK HAPPY BIRTHDAY BUNTING         3   
3124      536600     22411    JUMBO SHOPPER VINTAGE RED PAISLEY         6   
197768    553935     21090           SET/6 COLLAGE PAPER PLATES        12   
317093    564757     21471         STRAWBERRY RAFFIA FOOD COVER         3   
103303    545060     21519      GIN & TONIC DIET GREETING CARD         12   
474683    576845     23382  BOX OF 6 CHRISTMAS CAKE DECORATIONS         4   

             InvoiceDate  UnitPrice  CustomerID         Country  
536247    12/8/2011 9:28       4.96         NaN  United Kingdom  
227734   6/15/2011 13:05       8.25     16904.0  United Kingdom  
225964   6/14/2011 12:24       1.69     14730.0  United Kingdom  
446664   11/7/2011 16:16       1.65     17109.0  United Kingdom  
111799    3/7/2011 13:10       5.45     17634.0  United Kingdom  
3124      12/2/2010 8:32       1.65     17850.0  United Kingdom  
197768   5/20/2011 10:32       0.39     15856.0  United Kingdom  
317093   8/30/2011 10:36       3.29         NaN  United Kingdom  
103303   2/27/2011 13:47       0.42     15547.0  United Kingdom  
474683  11/16/2011 15:34       3.75     12991.0  United Kingdom

從抽樣的10行數據數據來看，數據符合「每列是一個變量，每行是一個觀察值，每個單元格是一個值」，具體來看每行是關於某商品的一次交易，每列是交易相關的各個變量，因此不存在結構性問題。

## 評估數據的乾淨度

In [5]:
orginal_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


從輸出結果來看，數據共有541909條觀察值，而 `Description`、`CustomerID` 變量存在缺失值。

此外，`InvoiceDate`的數據類型應為日期，`CustomerID` 的數據類型應為字串，應當進行數據格式轉換。

#### 評估缺失數據

在了解 `Description` 存在缺失值後，根據條件提取出缺失觀察值。

In [6]:
orginal_data[orginal_data["Description"].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]


有1454條交易數據缺失 `Description` 變量值。

從輸出結果來看，這些缺失 `Description` 的交易數據中，`UnitPrice` 都為0。為了驗證猜想，我們增加篩選條件，檢查是否存在 `Description`
變量缺失且 `UnitPrice` 不為0 的數據。

In [7]:
orginal_data[(orginal_data["Description"].isnull()) &(orginal_data["UnitPrice"] != 0) ]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []


篩選結果數量為0條，說明缺失 `Description` 值的數據，同時也不具備有效的 `UnitPrice` 值。

`Description` 表示產品名稱，`UnitPrice` 表示產品單價，都是進行後續商品交易分析的重要變量。如果它們同時缺失/無效，我們認為這些數據無法提供有效含義，因此可以在後續步驟中刪除。

`CustomerID`變量同樣存在缺失值，因此也根據條件提取出缺失觀察值。

In [8]:
orginal_data[orginal_data["CustomerID"].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

`CustomerID`表示客戶編號，並不是分析暢銷商品的必要變量。而且從輸出結果來看，有些`CustomerID`缺失的交易數據仍然有效，因此保留此變量為空的觀察值。

#### 評估重複數據


根據數據變量的含義來看，雖然`InvoiceNo`、`StockCode`和`CustomerID`都是唯一標識符，但一次交易可能包含多件商品，因此`InvoiceNo`可以存在重複；不同交易可以包含同件商品，因此 StockCode 可以存在重複；顧客可以進行多次交易或下單多個商品，因此`CustomerID`也可以存在重複。

因此，針對此數據集，我們無需評估重複數據。

#### 評估不一致數據

不一致數據可能存在於`Country`變量中，我們需要檢查是否存在多個不同值指代同一國家的情況。

In [9]:
orginal_data["Country"].value_counts()

United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68
European



從`Country` 變量的值來看，`"USA"` 和 `"United States"` 均表示美國，`"United States"`、`"UK"`、`"U.K."` 均表示英國，因此應在清洗步驟中對這些值進行統一處理，只保留一個指代值。

#### 評估無效或錯誤數據

可以通過 DataFrame 的 `describe`方法，快速了解數值型數據的統計信息。

In [10]:
orginal_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

從輸出結果來看，`Quantity` 和`UnitPrice`存在負數，可能會對後續的數值分析產生影響。

因此，我們先篩選出`Quantity`數值為負數的觀察值，進一步評估其含義。

In [11]:
orginal_data[orginal_data["Quantity"] < 0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

從篩選結果來看，`Quantity` 變量為負數的觀察值中，`InvoiceNo` 似乎均以 C 開頭，可能表示訂單被取消。

但此猜測不一定準確，為了驗證這一假設，我們可以增加篩選條件，檢查是否存在 `Quantity`變量為負數且 `InvoiceNo`不以 C 開頭的觀察值。

In [12]:
orginal_data[(orginal_data["Quantity"] < 0) & (orginal_data["InvoiceNo"].str[0]!= "C")]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

以上猜想被證實為錯誤，因為仍然存在 Quantity 變量為負數且 InvoiceNo 不以 C 開頭的觀察值。

但根據以上輸出結果，這些篩選出的觀察值的 UnitPrice 均為0，因此我們可以增加 UnitPrice 的條件進行進一步驗證。

In [13]:
orginal_data[(orginal_data["Quantity"] < 0) & (orginal_data["InvoiceNo"].str[0]!= "C")& (orginal_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

根據輸出結果，猜想得到驗證，當 `Quantity` 變量為負數時，觀察值滿足以下條件之一：

- `InvoiceNo` 以 C 開頭，表示訂單被取消
- `UnitPrice` 為0，說明單價為0英鎊
這些交易數據均不是有效成交數據，不貢獻銷售，因此不在後續分析範圍內。我們會在數據清理步驟中刪除 `Quantity`為負數的觀察值。

接下来篩選出`UnitPrice`為負數的觀察值。

In [14]:
orginal_data[orginal_data["UnitPrice"] < 0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

從輸出結果來看，`UnitPrice`為負數的觀察值都是壞賬調賬，不屬於實際商品交易數據，因此在數據清理步驟中也會將其刪除。

## 清理數據

根據前面的評估結論，我們需要進行的數據清理步驟包括：

- 把 `InvoiceDate` 變量的數據類型轉換為日期時間格式
- 把 `CustomerID` 變量的數據類型轉換為字串
- 把 `Description` 變量缺失的觀察值刪除
- 把 `Country `變量中值為 "USA" 替換為 "United States"
- 把 `Country `變量中值為 "UK" 和 "U.K." 替換為 "United Kingdom"
- 把 `Quantity `變量值為負數的觀察值刪除
- 把 `UnitPrice` 變量值為負數的觀察值刪除


為了區分經過清理的數據與原始數據，我們將創建一個新的變量 cleaned_data，它是從 original_data 複製出的副本，後續的清理步驟都將應用於 cleaned_data。

In [16]:
cleaned_data = orginal_data.copy()
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

將`InvoiceDate` 變量的數據類型轉換為日期時間類型：

In [17]:
cleaned_data["InvoiceDate"] = pd.to_datetime(cleaned_data["InvoiceDate"])
cleaned_data["InvoiceDate"]

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

將 `CustomerID`變量的數據類型轉換為字串：

In [18]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].astype(str)
cleaned_data["CustomerID"]

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

將 `CustomerID`值的結尾的".0"刪除：

In [19]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].str.slice(0, -2)
cleaned_data["CustomerID"]

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

將`Description`變量缺失的觀察值刪除，並查看刪除後該列的缺失值個數：

In [20]:
cleaned_data.dropna(subset=["Description"], inplace=True)

In [21]:
cleaned_data["Description"].isnull().sum()

0

將`Country`變量中值為`"USA"`的替換為 `"United States"`，並檢查替換後 `"USA"` 值的個數：

In [22]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"USA": "United States"})

In [23]:
len(cleaned_data[cleaned_data["Country"] == "USA"])

0

將`Country` 變量中值為` "UK"` 和` "U.K."` 的替換為` "United Kingdom"`，並檢查替換後 `"UK" `和 `"U.K."` 值的個數：

In [24]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"UK": "United Kingdom", "U.K.": "United Kingdom"})

In [25]:
print(len(cleaned_data[cleaned_data["Country"] == "UK"]))
print(len(cleaned_data[cleaned_data["Country"] == "U.K."]))

0
0


將 `Quantity` 變量中值為負數的觀察值刪除，並檢查刪除後 `Quantity` 變量中負數值的個數：

In [26]:
cleaned_data = cleaned_data[cleaned_data["Quantity"] >= 0]

In [27]:
len(cleaned_data[cleaned_data["Quantity"] < 0])

0

將 UnitPrice 變量中值為負數的觀察值刪除，並檢查刪除後 UnitPrice 變量中負數值的個數：

In [28]:
cleaned_data = cleaned_data[cleaned_data["UnitPrice"] >= 0]

In [29]:
len(cleaned_data[cleaned_data["UnitPrice"] < 0])

0

## 保存清理後的數據

完成數據清理後，把數據保存到新的文件名為`e_commerce_cleaned.csv`。

In [30]:
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice CustomerID         Country  
0 2010-12-01 08:26:00       2.55      17850  United Kingdom  
1 2010-12-01 08:26:00       3.39      17850  United Kingdom  
2 2010-12-01 08:26:00       2.75      17850  United Kingdom  
3 2010-12-01 08:26:00       3.39      17850  United Kingdom  
4 2010-12-01 08:26:00       3.39      17850  United Kingdom

In [31]:
cleaned_data.to_csv("e_commerce_cleaned.csv", index=False)

In [32]:
pd.read_csv("e_commerce_cleaned.csv").head()

/var/folders/d5/ltqm4csn30b6t48_h42p25gr0000gn/T/ipykernel_92696/2686214642.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv("e_commerce_cleaned.csv").head()


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice CustomerID         Country  
0  2010-12-01 08:26:00       2.55      17850  United Kingdom  
1  2010-12-01 08:26:00       3.39      17850  United Kingdom  
2  2010-12-01 08:26:00       2.75      17850  United Kingdom  
3  2010-12-01 08:26:00       3.39      17850  United Kingdom  
4  2010-12-01 08:26:00       3.39      17850  United Kingdom